code/model/modelreplace.cpp

```diff
@@ -0,0 +1,409 @@
+
+	toFill.name = name;
+	
+	required_string("+Base POF:");
+	stuff_string(toFill.basePOF, F_FILESPEC);
+
+	do {
+		ignore_white_space();
+		char operation_type[NAME_LENGTH];
+		stuff_string(operation_type, F_NAME, NAME_LENGTH);
+
+		auto operation_parser = virtual_pof_operations.find(operation_type);
+		if (operation_parser != virtual_pof_operations.end())
+			toFill.operationList.emplace_back(operation_parser->second());
+		else {
+			error_display(1, "Unknown operation type %s in virtual POF %s!", operation_type, name.c_str());
+			virtual_pof_list.pop_back();
+			skip_to_start_of_string_either("$POF:", "#End");
+			return;
+		}
+	} while (optional_string_either("$POF:", "#End") == -1);
+	Mp -= 5;
+}
+
+static void parse_virtual_pof_table(const char* filename) {
+	try
+	{
+		read_file_text(filename, CF_TYPE_TABLES);
+		reset_parse();
+
+		required_string("#Virtual POFs");
+
+		while (optional_string("$POF:"))
+			parse_virtual_pof();
+
+		required_string("#End");
+	}
+	catch (const parse::ParseException& e)
+	{
+		mprintf(("TABLES: Unable to parse '%s'!  Error message = %s.\n", filename, e.what()));
+		return;
+	}
+}
+
+void virtual_pof_init() {
+	if (cf_exists_full("virtual_pofs.tbl", CF_TYPE_TABLES))
+		parse_virtual_pof_table("virtual_pofs.tbl");
+
+	parse_modular_table(NOX("*-pof.tbm"), parse_virtual_pof_table);
+}
+
+// Internal helper functions
+
+#define REPLACE_IF_EQ(data, source, dest) if ((data) == source) (data) = dest;
+
+static void change_submodel_numbers(polymodel* pm, int source, int dest) {
+	//For now only in the subobject data...
+	//TODO Phase 2 expand to full polymodel
+	for (int i = 0; i < pm->n_models; i++) {
+		auto& submodel = pm->submodel[i];
+		for (auto& detail : submodel.details)
+			REPLACE_IF_EQ(detail, source, dest);
+		REPLACE_IF_EQ(submodel.first_child, source, dest);
+		REPLACE_IF_EQ(submodel.i_replace, source, dest);
+		for (auto& debris : submodel.live_debris)
+			REPLACE_IF_EQ(debris, source, dest);
+		REPLACE_IF_EQ(submodel.look_at_submodel, source, dest);
+		REPLACE_IF_EQ(submodel.my_replacement, source, dest);
+		REPLACE_IF_EQ(submodel.next_sibling, source, dest);
+		REPLACE_IF_EQ(submodel.parent, source, dest);
+	}
+}
+
+
+
+// Actual replacement operations
+
+VirtualPOFOperationAddSubmodel::VirtualPOFOperationAddSubmodel() {
+	required_string("+POF to Add:");
+	stuff_string(appendingPOF, F_FILESPEC);
+
+	required_string("+Source Subobject:");
+	stuff_string(subobjNameSrc, F_NAME);
+
+	if (optional_string("+Copy Children:")) {
+		stuff_boolean(&copyChildren);
+	}
+
+	if (optional_string("$Rename Subobjects:")) {
+		rename = make_unique<VirtualPOFOperationRenameSubobjects>();
+	}
+
+	required_string("+Destination Subobject:");
+	stuff_string(subobjNameDest, F_NAME);
+}
+
+void VirtualPOFOperationAddSubmodel::process(polymodel* pm, model_read_deferred_tasks& deferredTasks, model_parse_depth depth) const {
+	polymodel* appendingPM = new polymodel();
```

Comment:
Since you've used smart pointers elsewhere, why not make this a `unique_ptr`?

Reply:
Because we need to manually clear out it's dynamically allocated pointers using model_free anyways.
We could change model_free to not actually free the polymodel pointer itself and make it a smart pointer, but since it's still required to call model_free anyways, it'd defeat the purpose. To really make that worthwhile, we'd need to go through the model struct and replace all pointers and dynamically allocated memory with smart pointers (which might be a good idea, but that's out of scope for this PR, and is kinda needed all across the codebase).

Reply:
Ah, that makes sense.